airflow/contrib/kubernetes/worker_configuration.py

```diff
@@ -0,0 +1,174 @@
+            'mountPath': dags_path,
+            'readOnly': True
+        }]
+
+        # Mount the airflow.cfg file via a configmap the user has specified
+        if self.kube_config.airflow_configmap:
+            config_volume_name = "airflow-config"
+            config_path = '{}/airflow.cfg'.format(self.kube_config.airflow_home)
+            volumes.append({
+                'name': config_volume_name,
+                'configMap': {
+                    'name': self.kube_config.airflow_configmap
+                }
+            })
+            volume_mounts.append({
+                'name': config_volume_name,
+                'mountPath': config_path,
+                'subPath': 'airflow.cfg',
+                'readOnly': True
+            })
+
+        # A PV with the DAGs should be mounted
+        if self.kube_config.dags_volume_claim:
+            volumes[0]['persistentVolumeClaim'] = {
+                "claimName": self.kube_config.dags_volume_claim}
+            if self.kube_config.dags_volume_subpath:
+                volume_mounts[0]["subPath"] = self.kube_config.dags_volume_subpath
+        else:
+            # Create a Shared Volume for the Git-Sync module to populate
+            volumes[0]["emptyDir"] = {}
+        return volumes, volume_mounts
+
+    def _get_environment(self):
+        """Defines any necessary environment variables for the pod executor"""
+        env = {
+            'AIRFLOW__CORE__DAGS_FOLDER': '/tmp/dags',
+            'AIRFLOW__CORE__EXECUTOR': 'LocalExecutor'
+        }
+        if self.kube_config.airflow_configmap:
+            env['AIRFLOW__CORE__AIRFLOW_HOME'] = self.kube_config.airflow_home
+        return env
+
+    def _get_secrets(self):
+        """Defines any necessary secrets for the pod executor"""
+        worker_secrets = []
+        for env_var_name, obj_key_pair in six.iteritems(self.kube_config.kube_secrets):
+            k8s_secret_obj, k8s_secret_key = obj_key_pair.split('=')
+            worker_secrets.append(
+                Secret('env', env_var_name, k8s_secret_obj, k8s_secret_key))
+        return worker_secrets
+
+    def _get_image_pull_secrets(self):
+        """Extracts any image pull secrets for fetching container(s)"""
+        if not self.kube_config.image_pull_secrets:
+            return []
+        return self.kube_config.image_pull_secrets.split(',')
+
+    def make_pod(self, namespace, pod_id, dag_id, task_id, execution_date,
+                 airflow_command, kube_executor_config):
+        volumes, volume_mounts = self._get_volumes_and_mounts()
+        worker_init_container_spec = self._get_init_containers(
+            copy.deepcopy(volume_mounts))
+        resources = Resources(
+            request_memory=kube_executor_config.request_memory,
+            request_cpu=kube_executor_config.request_cpu,
+            limit_memory=kube_executor_config.limit_memory,
+            limit_cpu=kube_executor_config.limit_cpu
+        )
+
+        return Pod(
```

Comment:
Curious to know if you have direct log retrieval from the workers functioning?

One of the downsides of the traditional use of a K8s "Deployment" for the workers, is that DNS resolution Webserver -> Worker doesn't work, and you have to use storage buckets for logs - which means a bit of a wait for job logs.

Because you're using individual pod definitions, you can use the pod `hostname` and `subdomain` fields, combined with a headless service, to make DNS and worker log retrieval function properly.

Reply:
We were able to get direct real-time logging working for the kubernetes operator using the python-client (https://github.com/apache/incubator-airflow/commit/55f2674925d391e4f3930b24ca65d2cb65fc25c3)  so there's no reason we couldn't do the same for the executor. Could you please create a ticket on the apache airflow jira and link me to it? (If not then I will set one up later today).